File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-5.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {    
    public bool IsSubtree(TreeNode root, TreeNode subRoot) {
        //NOT A BINARY <<SEARCH>> TREE
        
        //Vacuous Case
        if(subRoot == null){
           return true;
        }

        //Reached the end of Base Tree
        if(root == null){
            return false;
        }

       
        //DFS PREORDER -> check on every node if same tree as subtree
        //######
        if(IsSameTree(root, subRoot)){
            return true;
        }

        return IsSubtree(root.left, subRoot) || IsSubtree(root.right, subRoot);
    }

   
    public bool IsSameTree(TreeNode p, TreeNode q){
        if(p == null && q == null){
            return true;
        }

        if(p == null || q == null || (p.val != q.val)){
            return false;
        }

        return IsSameTree(p.left, q.left) && IsSameTree(p.right, q.right);
    }

}

File: Data Structures & Algorithms/depth-of-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution: 
    def dfs(self, root, depth):
        if not root:
            return depth
        
        depth += 1

        return max(self.dfs(root.left, depth), self.dfs(root.right, depth))

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        return self.dfs(root, 0)